lib/Sema/MiscDiagnostics.cpp

```diff
@@ -818,15 +818,6 @@ static void diagSyntacticUseRestrictions(const Expr *E, const DeclContext *DC,
           return;
         }
       }
-      
-      // Unchecked casting to a subclass is better done by unsafeDowncast.
-      if (fromTy->isBindableToSuperclassOf(toTy)) {
```

Comment:
I read the description and the opting for removing the diagnostic all together, but I still wonder if it would be better to still emit this diagnostic when we and only when we know that we have both (from and to) concrete types. In that case we can know is `x is T` statically and safely emit the diagnostic and fix-it. 

That is normally how those runtime diagnostics warnings are emitted, only when compiler can know that information at compile time, so in the example, generic parameters there is no way to know that at compile time. 
To me it makes sense to just extend the check and still emit diagnostic when it can know for sure, but is just a personal opinion, since was stated in the ticket that removing diagnostic was also fine.

Reply:
+1 to leaving it in place, but restricting it to when we have sufficient information.

Reply:
Thank you @LucianoPAlmeida and @compnerd for the suggestions, It makes sense to do that. I will try to find way to apply it only we are sure. Thanks!

Reply:
It sounds like what we want here is to check that both sides are `ClassDecl` and that "from" class declaration is a subclass of "to" declaration via `ClassDecl::sSuperclassOf(ClassDecl *)`